include/librealsense2/h/rs_internal.h

```diff
@@ -49,6 +78,64 @@ rs2_context* rs2_create_recording_context(int api_version, const char* filename,
  */
 rs2_context* rs2_create_mock_context(int api_version, const char* filename, const char* section, rs2_error** error);
 
+/**
+ * Create software device to enable use librealsense logic without getting data from backend
+ * but inject the data from outside
+ * \param[out] error  if non-null, receives any error that occurs during this call, otherwise, errors are ignored
+ * \return            software device object, should be released by rs2_delete_device
+ */
+rs2_device* rs2_create_software_device(rs2_error** error);
+
+/**
+ * Add sensor to the software device
+ * \param[in] dev the software device
+ * \param[in] sensor_name the name of the sensor
+ * \param[out] error  if non-null, receives any error that occurs during this call, otherwise, errors are ignored
+ * \return            software sensor object, should be released by rs2_delete_sensor
+ */
+rs2_sensor* rs2_software_device_add_sensor(rs2_device* dev, const char* sensor_name, rs2_error** error);
+
+/**
+ * Inject frame to software sonsor
+ * \param[in] sensor the software sensor
+ * \param[in] frame all the frame components
+ * \param[out] error  if non-null, receives any error that occurs during this call, otherwise, errors are ignored
+ */
+void rs2_software_sensor_on_video_frame(rs2_sensor* sensor, rs2_software_video_frame frame, rs2_error** error);
+
+/**
+ * Set the wanted matcher type that will be used by the syncer
+ * \param[in] dev the software device
+ * \param[in] matcher matcher type
+ * \param[out] error  if non-null, receives any error that occurs during this call, otherwise, errors are ignored
+ */
+void rs2_software_device_create_matcher(rs2_device* dev, rs2_matchers matcher, rs2_error** error);
+
+/**
+ * Add video stream to sensor
+ * \param[in] sensor the software sensor
+ * \param[in] video_stream all the stream components
+ * \param[out] error  if non-null, receives any error that occurs during this call, otherwise, errors are ignored
+ */
+rs2_stream_profile* rs2_software_sensor_add_video_stream(rs2_sensor* sensor, rs2_video_stream video_stream, rs2_error** error);
+
+/**
+ * Add read only option to sensor
+ * \param[in] sensor the software sensor
+ * \param[in] option the wanted option
+ * \param[in] val the initial value
+ * \param[out] error  if non-null, receives any error that occurs during this call, otherwise, errors are ignored
+ */
+void rs2_software_sensor_add_read_only_option(rs2_sensor* sensor, rs2_option option, float val, rs2_error** error);
+
+/**
+ * Update the read only option added to sensor
+ * \param[in] sensor the software sensor
+ * \param[in] option the wanted option
+ * \param[in] val the wanted value
+ * \param[out] error  if non-null, receives any error that occurs during this call, otherwise, errors are ignored
+ */
+void rs2_software_sensor_update_read_only_option(rs2_sensor* sensor, rs2_option option, float val, rs2_error** error);
```

Comment:
why do you need 2 functions? can't we have a single set_option function?

Reply:
rs2_software_sensor_add_read_only_option is to register the option such that it will be supported by the sensor and give it some initial value,

and rs2_software_sensor_update_read_only_option is to update the value of registered option.